keps/sig-autoscaling/2556-circumvent-50-tag-asg-limit/README.md

```diff
@@ -0,0 +1,139 @@
+
+In the [AWS cloud provider][], if there are no nodes in the nodegroup, Cluster Autoscaler creates a nodeTemplate from ASG tags and some default allocatable resources. (See the [code][]) The tags for Cluster Autoscaler include resources, labels, and taints. The nodeTemplate is used to identify available node resources in the absence of an actual EC2 instance existing. We propose that, when a scaled-to-0 EKS ManagedNodegroup is being used, the Cluster Autoscaler cloud provider for AWS takes advantage of the EKS DescribeNodegroup API, which will return the latest lists of labels and taints. Unlike the ASG tags, the API will not limit the number of labels and taints that Cluster Autoscaler can discover to 50.
+
+[AWS cloud provider]: https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/aws
+[code]: https://github.com/kubernetes/autoscaler/blob/99936f3f3971c91b30def913c1c807c42c4b87a7/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L365-L371
+
+## IAM
+
+The ServiceAccount running the Cluster Autoscaler Pod (if using IAM Roles for Service Accounts (IRSA)) or associated with the instance profile (if not using IRSA) will need to have one additional policy applied to it: `eks:DescribeNodegroup`. If Cluster Autoscaler doesn’t have the correct permissions to call the API, it will fall back to using just the ASG tags.
+
+## Design
+
+We propose that the AWS specific driver of Cluster Autoscaler checks for AWS EKS ManagedNodegroups tags on the ASG that marks the ASG as a ManagedNodegroup. If the tags are present, Cluster Autoscaler pulls information from a cache struct containing DescribeNodegroup API response data. The AWS EKS ManagedNodegroups tags look like  `eks:cluster-name : <CLUSTER_NAME>` and `eks:nodegroup-name : <NODEGROUP_NAME>`. We'll use AWS EKS ManagedNodegroups tags because we automatically add them to every managed nodegroup. The DescribeNodegroup API is already in the AWS SDK that Cluster Autoscaler uses. ([Current EKS interface][])
+
+When Cluster Autoscaler finds the ManagedNodegroups tags, it will call functions on a ManagedNodegroupCache struct to get the labels, taints, and some other values. The cache struct will hold the response data from the EKS [DescribeNodegroup API][]. The cache struct methods will call the DescribeNodegroup API when the cached data’s TTL (1 minute default, but will be configurable) has expired or there’s no cached information for the ManagedNodegroup. Cluster Autoscaler will include both the ASG tag values and the values from the EKS API in its decisions. If an ASG tag contains the same resource as the EKS API (ex: a label key is used in both), Cluster Autoscaler will choose the value from the ASG tag so customers can override any value they want.
```

Comment:
This currently seems a bit unclear when first read as to whether the cache structure will automatically rehydrate the cache when a cache entry's TTL has expired in the background, or whether it will only happen when the cache is accessed and the TTL is found to have expired.

Reply:
I'll clarify. Currently the plan is to only refresh when the cache is accessed and the TTL is found to have expired.